feat(cmd): add update command

Malet opened the pull request:
- Adds `corectl update` command
- Automatically checks for updates every hour when other commands are invoked (except the update command, and for completions)
- Updates during other commands will run the original command after the update has been completed
- Deprecated `--nonint` in favour of `--non-interactive` flag
- Show warning when config is not initialised, only enable commands that require config once it's initialised
- Auto-update (during other commands) is disabled when using `--non-interactive` flag
- Changed `--non-interactive` to now be a global flag
- Add `Abort` behaviour for `wizard`, for user feedback when a task fails
- Fixed <kbd>Ctrl</kbd>+<kbd>C</kbd> behaviour of `wizard` with injected input component
- Created confirmation prompt component
- Add local integration test suite for commands which don't require config setup and github auth
- Fix nil pointer bug when multiple wizard instances are used (correctly unset the current handler) 

![term](https://github.com/user-attachments/assets/ef6f04c1-aa6e-4a6b-a6b0-9967f3416b1c)

Malet committed:
feat(cmd): add update command

Malet committed:
feat(update): skip update if you're already up-to-date!

Malet committed:
feat(tests): add unit tests for github release helpers

Malet committed:
feat(tests): implement e2e test for update functionality

feat(update): improve output display for update, add fail function for
wizard

Malet committed:
fix(tests): use exisitng corectl config test setup

Malet committed:
feat(tests): add local integration test suite

feat(config): allow commands to be run without config where possible

Malet committed:
feat(update): improve update compatibility

Malet committed:
feat(tests): copy of prebuilt corectl, create tmp copy for update test

Malet committed:
feat(tests): add local integration tests to makefile

Malet committed:
fix(lint): ensure all errors are checked

Malet committed:
feat(update): add confirmation dialog

Malet committed:
fix(nonint): deprecate --nonint, update to --non-interactive

This creates a global flag which is used by all subcommands

Malet committed:
feat(update): run auto update in all other commands every hour

Malet committed:
feat(refactor): clean up confirmation component styles

Malet committed:
fix(update): don't auto update when running update command

Malet committed:
fix(tests): update integration tests to add --non-interactive

Malet committed:
fix(update): fix completions (do not run update or configure logs)

Malet committed:
fix(git): fix items lost in rebase

Mugenor commented on pkg/cmdutil/shell/shell.go:
```diff
@@ -0,0 +1,85 @@
+	"github.com/phuslu/log"
+)
+
+func MoveFile(source string, destination string) error {
```
Isn't this file a copy of os/util.go?

Malet replied:
Yep, I renamed it to avoid the namespace clash, removed!

Mugenor replied:
Have you pushed it?
I still can see the duplicated file

Malet replied:
I have now

Mugenor commented on pkg/cmdutil/shell/shell.go:
```diff
@@ -0,0 +1,85 @@
+	return nil
+}
+
+func CopyFile(source string, destination string) error {
```
We a dependency for the lib that can copy files, we could use it: `copy.Copy()`

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,286 @@
+	updateInterval := 1 * time.Hour
+	updateStatusFileName := "corectl-autoupdate"
+	log.Debug().Msg("checking for updates")
+	nonInteractive, err := cmd.Flags().GetBool("non-interactive")
```
I would propose to also try to deduce this flag automatically by checking if we are working with a terminal or not.
Similar to what we do here: https://github.com/coreeng/corectl/blob/829d30c28592d75ac707ef496359e7450cd5fb87/pkg/cmdutil/userio/prompt.go#L65
This flag was meant to be just as an option to explicitly disable all interactivity

Malet commented on pkg/cmd/root/root.go:
```diff
@@ -33,18 +38,33 @@ func ConfigureGlobalLogger(logLevelFlag string) {
+
 func NewRootCmd(cfg *config.Config) *cobra.Command {
 	var logLevel string
+	var nonInteractive bool
```
I'm using it via `nonInteractive, err := cmd.Flags().GetBool("non-interactive")`. We have to give it a variable to bind to, even if we don't use it directly

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,286 @@
+	timeSince := now.Sub(previousTime)
+	if timeSince >= updateInterval {
+		githubToken := ""
+		if cfg.IsPersisted() {
```
I think you can just directly get the value from the config, without this check. In the worst case it will be an empty string. Otherwise it's a value from the config file or from the passed flag (can it be the case?)

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,286 @@
+		release, err = git.GetCorectlReleaseByTag(githubClient, opts.targetVersion)
+	}
+	if err != nil {
+		wizard.Abort(err)
```
We shouldn't panic when the error is caused by any external means, we should handle it somehow or propagate to the caller

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,286 @@
+	if err != nil {
+		wizard.Abort(err)
+		log.Warn().Err(err).Msg("could not render changelog markdown, falling back to plaintext")
+		fmt.Print(asset.Changelog)
```
we should use `IOStreams` object for printing things

Mugenor replied:
Please, check all other places.
There are usages in root.go, update.go and version.go files

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,286 @@
+			wizard.Info("Update accepted")
+		} else {
+			wizard.Abort(fmt.Errorf("Update cancelled by user"))
+			wizard.Info("Update cancelled")
```
`Abort` will kill wizard, so `Info` shouldn't work, should it?

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,286 @@
+		} else {
+			wizard.Abort(fmt.Errorf("Update cancelled by user"))
+			wizard.Info("Update cancelled")
+			return
```
Will it ask user to update every time user refuses to do it?
I would say it can be very annoying

Malet committed:
fix(wizard): use new TaskStatus for wizard.Abort implementation

Malet committed:
feat(wizard): remove redundant failed flag

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,286 @@
+			return
+		}
+
+		update(UpdateOpts{
```
I would propose to just give user a warning/info message about a new version instead of asking him to download the new release right in the middle of another operation

Mugenor commented on pkg/cmd/root/root.go:
```diff
@@ -63,21 +83,55 @@ func NewRootCmd(cfg *config.Config) *cobra.Command {
-	if err != nil {
-		panic("Unable to execute p2p command")
+
+	if cfg.IsPersisted() {
```
Do we really want to do it?
I think it should not be necessary to run the initialization on CI, for example. 
It should be possible to simply specify everything corectl needs via flags

Malet replied:
it currently *is* necessary to run init on CI, these commands already exit if the config isn't set

Mugenor replied:
Can you give an example?
It shouldn't be the case, unless you didn't set up any of the necessary flags

Malet replied:
```
➜  corectl git:(feature/update) ✗ ./corectl tenant create --dry-run
...
Error: cplatform repository path is not set. consider initializing corectl first:
  corectl config init
```

```
➜  corectl git:(feature/update) ✗ ./corectl app create test --tenant michael --dry-run
INFO: resetting local repository [repo=]
INFO: resetting local repository [repo=]
📝 Creating new application test: https://github.com//test
                                         
  1. <empty>                             
  2. blank                               
> 3. go-web                              
...

Error: couldn't load tenant configuration in path tenants/tenants: stat .: no such file or directory
```

```
➜  corectl git:(feature/update) ✗ ./corectl p2p export --tenant michael -e gcp-dev
Usage:
  corectl p2p export [flags]
...
Error: cplatform repository path is not set. consider initializing corectl first:
  corectl config init
```

I can see now that you *could* set these via the params (I don't know if this is very likely outside of scripts), I'll roll this back

Mugenor commented on pkg/cmd/application/create/app_create.go:
```diff
@@ -60,10 +61,16 @@ NOTE:
 				opts.LocalPath = "./" + opts.Name
 			}
+
+			nonInteractive, err := cmd.Flags().GetBool("non-interactive")
```
What do you think about creating a struct for global objects and passing it together with config struct? Or create a wrapper struct to pass all of it as a single arg?

Malet replied:
I'd rather leave it as-is for now, this PR is already huge :laughing:

Mugenor commented on pkg/cmd/root/root.go:
```diff
@@ -33,18 +38,33 @@ func ConfigureGlobalLogger(logLevelFlag string) {
-			cmd.SilenceErrors = true
+			if !isCompletion() {
+				ConfigureGlobalLogger(logLevel)
+				if cmd.Name() != "update" {
```
Let's extract it to a global constant in cmd/update.go and reuse it here and there

Malet replied:
it's only needed in root, we haven't actually called the cmd at this point

Mugenor replied:
I don't think I understand your point.
You don't have to call anything to use a constant

Mugenor commented on pkg/cmd/root/root.go:
```diff
@@ -63,21 +83,55 @@ func NewRootCmd(cfg *config.Config) *cobra.Command {
+		if !isCompletion() {
+			styles := userio.NewNonInteractiveStyles()
+			fmt.Println(
+				styles.WarnMessageStyle.Render("Config not initialised, please run ") +
```
It will be also printed in case you run an init command, won't it?

Mugenor commented on pkg/cmdutil/shell/shell.go:
```diff
@@ -0,0 +1,85 @@
+	"github.com/phuslu/log"
+)
+
+func MoveFile(source string, destination string) error {
```
I would also propose to move MoveFile and RunCommand functions right to file where they are used, since they are rather small and used only in one place. I also don't think have such util packages is a good idea, since they can bloat with random stuff over time (we have this problem in developer-platform functional tests).
But I don't insist here

Mugenor commented on pkg/cmdutil/userio/wizard/wizard.go:
```diff
@@ -131,12 +135,29 @@ func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+			return m, tea.Sequence(
+				func() tea.Msg {
+					return updateCurrentTaskCompletedTitle{
+						title:  msg.Error(),
```
Suggestion: if we use an error just to display a message, we can pass string directly, without wrapping it as an error. I think it's more convenient

Mugenor commented on pkg/cmdutil/userio/wizard/wizard.go:
```diff
@@ -177,21 +198,18 @@ func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 		log.Debug().Msgf("Wizard: Received keystroke [%s]", msg.String())
-
 		switch msg.Type {
 		case tea.KeyCtrlC:
```
This will pass keystroke to the input model only in ctrl+c is entered, won't it? Or am I missing something?

Mugenor replied:
Ah, I didn't notice you pass every message to the inner model a bit later, sorry

Mugenor commented on pkg/cmdutil/userio/wizard/wizard.go:
```diff
@@ -177,21 +198,18 @@ func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 				return m, inputCmd
 			}
+			m.quitting = true
+			return m, tea.Quit
```
Suggestion: should we pass a doneMsg instead, so the logic with completing all the hanging messages will also execute?

Malet replied:
For a <kbd>ctrl</kbd>+<kbd>c</kbd> I think we just want to exit regardless of anything in the message queue

Mugenor commented on pkg/cmdutil/userio/confirmation/confirmation.go:
```diff
@@ -0,0 +1,160 @@
+func GetInput(streams userio.IOStreams, question string) (bool, error) {
+	modelInstance := newModel(question)
+
+	result, err := streams.Execute(modelInstance, nil)
```
I changed IOStreams interface to accept tea.ProgramOptions, so you should remove nil argument

Mugenor commented on pkg/cmdutil/userio/interop.go:
```diff
@@ -12,15 +13,16 @@ type nonInteractiveHandler struct {
 func InfoLog(message string) string {
-	style := newNonInteractiveStyles()
-	return fmt.Sprintf("%s %s", style.infoStyle.Render("INFO:"), message)
+	style := NewNonInteractiveStyles()
```
I've added `styles` field to the struct, so you can reuse it instead of reinitializing it. Same for `WarnLog`

Mugenor commented on pkg/cmdutil/userio/prompt.go:
```diff
@@ -77,18 +77,21 @@ func isTerminalInteractive(in io.Reader, out io.Writer) bool {
 			model,
 			options...,
 		).Run()
+		s.CurrentHandler = nil
```
What is the reason for setting it to `nil` here?

Malet replied:
once the currenthandler finishes we want new requests to create their own program instead of try to reuse an existing one (which will no longer exist)

Mugenor replied:
But how does it help?
This code is execute in two cases:
- CurrentHandler is already nil
- The passed model is wizard. In this case handler will be updated in the caller function

Mugenor commented on pkg/cmdutil/userio/styles.go:
```diff
@@ -42,19 +42,19 @@ func newStyles(renderer *lipgloss.Renderer) *styles {
-	warnMessageStyle lipgloss.Style
-	bold             lipgloss.Style
-	status           wizard.TaskStatusStyle
+	InfoStyle        lipgloss.Style
```
Should we make the struct also public, since the constructor func and all the fields are also public?

Malet replied:
we don't technically need to since you can get it from `userio.NewNonInteractiveStyles()`

Mugenor replied:
It just feels very wrong :D
You return to the outer world unexported type and say that all of it fields are exported for some reason
It's kind of minor, but I think it brings some confusion. I wouldn't expect to see private type usages in other packages

Malet replied:
updated :+1:

Mugenor commented on pkg/git/github.go:
```diff
@@ -49,6 +57,119 @@ func DeriveRepositoryFullnameFromUrl(githubRepoUrl string) (RepositoryFullname,
 	}, nil
 }
 
+type CoreCtlAsset struct {
```
I would say we shouldn't keep this logic in `git` package and we should move near the update cmd, since it's very logic to corectl udpate

Mugenor commented on pkg/git/github.go:
```diff
@@ -49,6 +57,119 @@ func DeriveRepositoryFullnameFromUrl(githubRepoUrl string) (RepositoryFullname,
+func GetLatestCorectlAsset(release *github.RepositoryRelease) (*CoreCtlAsset, error) {
+	dummyAsset := CoreCtlAsset{}
+	if release.Assets == nil {
+		return &dummyAsset, errors.New("no assets found for the latest release")
```
Minor: why not simply returning nil?

Mugenor commented on tests/integration/suite_test.go:
```diff
@@ -19,12 +19,12 @@ import (
 	"github.com/thanhpk/randstr"
 
 	// Test cases import
+	_ "github.com/coreeng/corectl/pkg/cmd/update"
```
Probably a typo

Mugenor commented on tests/localintegration/suite_test.go:
```diff
@@ -0,0 +1,19 @@
+)
+
+// For any commands which do not need write access to github apis or local config to be initialised
+func TestLocalSuite(t *testing.T) {
```
Does it also run on CI?

Malet replied:
yep, through the makefile: `integration-test: build integration-test-local`

Mugenor commented on tests/localintegration/update/update.go:
```diff
@@ -0,0 +1,72 @@
+	Context("from local build", func() {
+		It("updates the version to latest", func() {
+			initialVersion, updatedVersion, err := updateCmd([]string{})
+			if err != nil {
```
Expect(err).NotTo(HaveOccurred())

Mugenor commented on tests/localintegration/update/update.go:
```diff
@@ -0,0 +1,72 @@
+
+		Expect(err).ShouldNot(HaveOccurred())
+		tmpPath, err := os.Readlink(fmt.Sprintf("/proc/self/fd/%d", tmpFile.Fd()))
+		tmpFile.Close()
```
Does it have to be closed?

Malet committed:
fix(review): incorporate PR feedback

Malet committed:
fix(update): disable auto update when tty is noninteractive

Malet committed:
fix(handler): add warning if handler is closed multiple times

fix(tests): use idiomatic syntax for errors
fix(update): remove redundant config check

Malet committed:
feat(refactor): dissolve utils into where they are used

Malet committed:
fix(confirmation): correct execute params

Malet committed:
fix(update): check errors for changelog print

Malet committed:
feat(refactor): move update code out of github module

Malet committed:
fix(tests): move tests from github to update module

Malet committed:
fix(wizard): set task status to success when marking complete

Malet committed:
fix(config): only show config warning if there's no file persisted

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,453 @@
+		err = fmt.Errorf("unable to create temporary file %s: %+v", asset.Version, err)
+		return err
+	}
+	tmpPath, err := os.Readlink(fmt.Sprintf("/proc/self/fd/%d", tmpFile.Fd()))
```
tested it on my mac and it doesn't work

Malet replied:
I'll test it out on mine and find an alternative way

Malet replied:
<img width="787" alt="Screenshot 2024-10-22 at 13 48 31" src="https://github.com/user-attachments/assets/8138357e-3ddf-497e-9e2f-e4e14b9008b2">
Working now

Mugenor commented on pkg/cmd/root/root.go:
```diff
@@ -33,18 +38,41 @@ func ConfigureGlobalLogger(logLevelFlag string) {
 }
 
+func isCompletion() bool {
+	return len(os.Args) >= 1 && os.Args[1] == "__complete"
```
it panics if I run just `corectl`

Malet replied:
out by 1 🤣 fixed

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,453 @@
+				wizard.Info("Update accepted")
+			} else {
+				err = fmt.Errorf("update cancelled by user")
+				wizard.Abort(err.Error())
```
it panics if I do not confirm the update

Malet replied:
panic is due to the other check we added: `{"time":"2024-10-22T13:55:00.948+01:00","level":"panic","message":"handler is already completed"}` will fix!

Mugenor commented on pkg/cmd/update/update.go:
```diff
@@ -0,0 +1,453 @@
+				opts.targetVersion = args[0]
+			}
+			if cfg.IsPersisted() {
+				opts.githubToken = cfg.GitHub.Token.Value
```
Let's add a flag for github token and not check if token is persisted

Malet replied:
the token is optional for the release endpoints, I'll remove the IsPersisted though

Malet committed:
fix(root): fix out by one error in isCompletion

Malet committed:
fix(update): use different method to get temp file name to support MacOS

Malet committed:
feat(styles): make NonInteractiveStyles public

Malet committed:
fix(handler): fixes for handler.Abort

Malet committed:
fix(cfg): don't check if config is persisted

Malet committed:
fix(wizard): fix current handler behaviour

fix(update): fix Done being called via defer even after Abort

Malet committed:
fix(stdout): use ioStreams for all printing

fix(strings): use constant instead of magic string for update and config cmd check

Malet committed:
feat(iostream): implement log warn and error levels, normalise
